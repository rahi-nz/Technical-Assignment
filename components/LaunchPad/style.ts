import styled from "styled-components";

export const Container = styled.section`
  padding: 20px;
  text-align: left;
  width: 50%;
`;

export const Lists = styled.div`
  background: #bde2e8;
  padding: 10px;
`;

export const Title = styled.p`
  font-size: 18px;
`;
