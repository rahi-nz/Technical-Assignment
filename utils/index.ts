export const failedLaunches = (name, launches) => ({
  launchpad: name,
  all_failures: launches
    .filter((el) => el.data.success === false)
    .map((el) => ({
      name: el.data.name,
      failures: el.data.failures[0].reason,
    })),
});

const months = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December",
];

export const launchedDate = (data) => {
  const res = data.map((el) => {
    const arr = el.date.split("-");
    return `starlinks id:${el.id} launched on ${
      months[Number(arr[1])]
    } ${Number(arr[2])}th ${Number(arr[0])}`;
  });
  return res;
};
