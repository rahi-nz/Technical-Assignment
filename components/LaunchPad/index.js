import React, { useEffect, useState } from "react";
import { request } from "../../store/request";
import { getLaunchpad, getLaunches } from "../../resources/Api";
import { failedLaunches } from "../../utils";
import { launchpadID } from "../../webConfig";
import { Container, Lists, Title } from "./style";

const LaunchPad = () => {
  const [failed, setFailed] = useState();
  const getLaunchesData = async (value) =>
    Promise.all(value.map((item) => request.get(getLaunches(item))));

  useEffect(() => {
    (async () => {
      const response = await request.get(getLaunchpad(launchpadID));
      if (response.ok) {
        const launches = await getLaunchesData(response.data.launches);
        const failedLaunch = failedLaunches(response.data.name, launches);
        setFailed(failedLaunch);
      }
    })();
    }, []); // eslint-disable-line

  return (
    <Container>
      <h3>FailedLaunches</h3>
      {failed && (
        <Lists>
          <Title>{failed?.name}</Title>
          {failed?.all_failures.map((el) => (
            <ul key={el.name}>
              <li>name: {el.name}</li>
              <li>reason: {el.failures}</li>
            </ul>
          ))}
        </Lists>
      )}
    </Container>
  );
};

export default LaunchPad;
