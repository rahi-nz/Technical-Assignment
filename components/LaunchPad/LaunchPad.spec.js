import React from "react";
import createStore from "redux-mock-store";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import LaunchPad from "./index";
import { failedLaunches } from "../../utils";

describe("<LaunchPad />", () => {
  let rootElement;
  let testRenderer;
  let store;
  let data;
  let result;

  beforeEach(async () => {
    result = {
      launchpad: "Kwajalein Atoll",
      all_failures: [
        {
          name: "FalconSat",
          failures: "merlin engine failure",
        },
      ],
    };
    data = [
      {
        config: {},
        data: {
          auto_update: true,
          capsules: [],
          cores: [{}],
          crew: [],
          date_local: "2006-03-25T10:30:00+12:00",
          date_precision: "hour",
          date_unix: 1143239400,
          date_utc: "2006-03-24T22:30:00.000Z",
          details: "Engine failure at 33 seconds and loss of vehicle",
          failures: [
            {
              altitude: null,
              reason: "merlin engine failure",
              time: 33,
            },
          ],
          fairings: {
            reused: false,
            recovery_attempt: false,
            recovered: false,
            ships: Array(0),
          },
          flight_number: 1,
          id: "5eb87cd9ffd86e000604b32a",
          launchpad: "5e9e4502f5090995de566f86",
          links: {},
          name: "FalconSat",
          net: false,
          payloads: ["5eb0e4b5b6c3bb0006eeb1e1"],
          rocket: "5e9d0d95eda69955f709d1eb",
          ships: [],
          static_fire_date_unix: 1142553600,
          static_fire_date_utc: "2006-03-17T00:00:00.000Z",
          success: false,
          tbd: false,
          upcoming: false,
          window: 0,
        },
        duration: 538,
        headers: {},
        ok: true,
        originalError: null,
        problem: null,
        status: 200,
      },
    ];
    store = createStore([])({
      satellites: {
        loading: false,
        data: [],
        error: false,
      },
    });
    store.dispatch = jest.fn();

    rootElement = () => (
      <Provider store={store}>
        <LaunchPad />
      </Provider>
    );

    testRenderer = renderer.create(rootElement());
  });

  it("Should return failed Launches", () => {
    expect(failedLaunches("Kwajalein Atoll", data)).toEqual(result);
  });

  it("Should render correctly", () => {
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
