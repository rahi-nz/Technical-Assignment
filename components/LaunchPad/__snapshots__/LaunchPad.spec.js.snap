// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaunchPad /> Should render correctly 1`] = `
<section
  className="style__Container-sc-1hovddl-0 enkUkG"
>
  <h3>
    FailedLaunches
  </h3>
</section>
`;
